### DISCUSSION TITLE
How can I configure {d,v,c}af to target preceding comments and whitespace?

### DISCUSSION THREAD

**Discussion by ColinKennedy** [nvim-treesitter/nvim-treesitter-textobjects#452] [Q&A]
I use nvim-treesitter-context for Python and C++ development. While it is useful, one property of other plug-ins which aim to do the same thing that I miss very much is for motions such as `daf` to effectively mean "Once executed, your cursor is just one `o` away from making a new function as if the previous function never existed". It's easier to explain visually.

`|X|` is the current cursor position

```python
class Foo2():
    def __init__(self):
        super().__init__()


    # some comment related to ``another``
    def another(self):
        |X|print("more")



    def last(self):
        print("more")
```

When I press `daf`, I want this output

```python
class Foo2():
    def __init__(self):
        super().__init__()


|X|


    def last(self):
        print("more")
```

Now if I press `o`, the cursor is exactly where it should be define a new function.

```python
class Foo2():
    def __init__(self):
        super().__init__()


    |X|



    def last(self):
        print("more")
```

And the pre / post whitespace around the methods have been preserved (2 lines above, 3 lines below)

Instead though, I get this

```python
class Foo2():
    def __init__(self):
        super().__init__()


    # some comment related to ``another``   <-- I wanted `daf` to remove this comment as well
    |X|



    def last(self):
        print("more")
```

I tried modifying [queries/python/textobjects.scm](https://github.com/nvim-treesitter/nvim-treesitter-textobjects/blob/a955ea4291a715ad72d63be4bd3bb5a4f9320426/queries/python/textobjects.scm) to capture preceding comments as well like this

```scm
(
  (comment)*
  (function_definition
    body: (block)?) @function.inner
) @function.outer
```

And a few similar variations but none of them work. Any advice on what query to use would be appreciated. Thank you!

**Comment by theHamsta** [comment]
White spaces can be handled by this option. https://github.com/nvim-treesitter/nvim-treesitter-textobjects/blob/35a60f093fa15a303874975f963428a5cd24e4a0/README.md#L79 This currently only includes white spaces and not comments, but a similar option could be added (would require a PR).

A problem with tree-sitter queries is that they always just can select single nodes. We found a work-around with `@function.outer.start`. This would only work for a single comment. See how C++ optionally includes `template <typename T>` to a function https://github.com/nvim-treesitter/nvim-treesitter-textobjects/blob/35a60f093fa15a303874975f963428a5cd24e4a0/queries/cpp/textobjects.scm#L12-L13 . There is also the more flexible `#make-range!` predicate that will likely be removed in the refactoring nvim-treesitter and maybe replaced by something else.

To select all nodes of `(comment)*` Neovim would need that PR: https://github.com/neovim/neovim/pull/17099

**Reply by ColinKennedy** [reply]
Just so I understand, are you saying that even if https://github.com/neovim/neovim/pull/17099 is merged, what I was describing still would not work due to a tree-sitter query limitation or that it should be possible after that PR is merged? It sounded like you were suggesting the latter but I wasn't quite sure from your response. `include_surrounding_whitespace` isn't really the right effect but I appreciate that you're trying to provide options, thank you :)

Thinking aloud for a moment - it sounded like if https://github.com/neovim/neovim/pull/17099 is merged is merged then we could add optional "delete comments" support via a new capture group like `(comment)* @comment`, explicitly or by reading the `@comment` that tends to come from parsers already installed via nvim-treesitter. And then, with a new PR to nvim-treesitter-textobjects, one could have a configuration option like `trim_preceding_comments = true` or something to that effect which would look for `@comment` info and delete the nodes associated with it. Is that on the right track?
